plumbing/transport/file/server_test.go

```diff
@@ -0,0 +1,109 @@
+package file
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+
+	"gopkg.in/src-d/go-git.v4/fixtures"
+
+	. "gopkg.in/check.v1"
+)
+
+type ServerSuite struct {
+	CommonSuite
+	RemoteName string
+	SrcPath    string
+	DstPath    string
+	DstURL     string
+}
+
+var _ = Suite(&ServerSuite{})
+
+func (s *ServerSuite) SetUpSuite(c *C) {
+	s.CommonSuite.SetUpSuite(c)
+
+	s.RemoteName = "test"
+
+	fixture := fixtures.Basic().One()
+	s.SrcPath = fixture.DotGit().Base()
+
+	fixture = fixtures.ByTag("empty").One()
+	s.DstPath = fixture.DotGit().Base()
+	s.DstURL = fmt.Sprintf("file://%s", s.DstPath)
+
+	cmd := exec.Command("git", "remote", "add", s.RemoteName, s.DstURL)
+	cmd.Dir = s.SrcPath
+	c.Assert(cmd.Run(), IsNil)
+}
+
+func (s *ServerSuite) TestPush(c *C) {
+	// git <2.0 cannot push to an empty repository without a refspec.
+	cmd := exec.Command("git", "push",
+		"--receive-pack", s.ReceivePackBin,
+		s.RemoteName, "refs/heads/*:refs/heads/*",
+	)
+	cmd.Dir = s.SrcPath
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "GIT_TRACE=true", "GIT_TRACE_PACKET=true")
+	stdout, stderr, err := execAndGetOutput(c, cmd)
+	c.Assert(err, IsNil, Commentf("STDOUT:\n%s\nSTDERR:\n%s\n", stdout, stderr))
+}
+
+func (s *ServerSuite) TestClone(c *C) {
+	pathToClone := c.MkDir()
+
+	cmd := exec.Command("git", "clone",
+		"--upload-pack", s.UploadPackBin,
+		s.SrcPath, pathToClone,
+	)
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "GIT_TRACE=true", "GIT_TRACE_PACKET=true")
+	stdout, stderr, err := execAndGetOutput(c, cmd)
+	c.Assert(err, IsNil, Commentf("STDOUT:\n%s\nSTDERR:\n%s\n", stdout, stderr))
+}
+
+func execAndGetOutput(c *C, cmd *exec.Cmd) (stdout, stderr string, err error) {
+	sout, err := cmd.StdoutPipe()
+	c.Assert(err, IsNil)
+	serr, err := cmd.StderrPipe()
+	c.Assert(err, IsNil)
+
+	outChan := make(chan string, 1)
+	outDoneChan := make(chan error, 1)
+	errChan := make(chan string, 1)
+	errDoneChan := make(chan error, 1)
+
+	c.Assert(cmd.Start(), IsNil)
+
+	go func() {
+		b, err := ioutil.ReadAll(sout)
+		if err != nil {
+			outDoneChan <- err
+		}
+
+		outChan <- string(b)
+		outDoneChan <- nil
```

Comment:
I've moved this code to a function, which should be more legible. Also changed the error channel to always return a value (either error or nil), just as if it was a regular error return value.